### PR TITLE
feat: add algorithm for length of longest substring w/o repeating characters

### PR DESCRIPTION
**Pull request by nishant-mahesh**
#### Description of Change
This cpp file contains an original implementation of an algorithm to solve the problem of finding the **longest substring of a given string, such that no character occurs more than once**. Few simple testcases have been hardcoded for checking purposes.

#### Checklist
- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Solution for issue: **[OTHER] Add Longest substring (without repeating characters) implementation #1672**

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1870"></a>

### CODE REVIEW

**Comment by nishant-mahesh** [comment]
Hello @Panquesito7! This is my first pull request, and I am just getting introduced to open source contribution. Could you let me know if I've done things right with this pull request? 

This pull request deals with [this issue](https://github.com/TheAlgorithms/C-Plus-Plus/issues/1672),

**Comment by Panquesito7** [comment]
> Hello @Panquesito7! This is my first pull request, and I am just getting introduced to open source contribution. Could you let me know if I've done things right with this pull request?
> 
> This pull request deals with [this issue](https://github.com/TheAlgorithms/C-Plus-Plus/issues/1672).

Hey there! For sure! I'll help you with this PR. Thanks for contributing! 😃

File: strings/longest_substr_without_repeating_chars.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/**
+ * @file
+ * @brief Greedy algorithm to find length of longest substring without repeating
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: strings/longest_substr_without_repeating_chars.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * @author [Nishant Mahesh](https://github.com/nishant-mahesh)
+ */
+
+#include <algorithm>  // To use the inbuilt 'max' function
+#include <cassert>
+#include <iostream>
+#include <map>
+#include <string>
+
+using namespace std;
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:22]: This is not good practice. It is best to use `std::` directly.
Or you can do this instead. This is good practice.

```suggestion
using std::cout;
using std::cin;
using std::string;
```

You can add more using sentences as well. 🙂

File: strings/longest_substr_without_repeating_chars.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * This function takes a string as input, and returns the length of the longest
+ * substring that contains no repeated characters
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:26]: Missing parameter documentation.

File: strings/longest_substr_without_repeating_chars.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        } else {                           // no repetitions encoutnered
+            cur_len += 1;
+            prev_occurence[input[i]] = i;  // update the prev_occrence map
+        }
+    }
+    ans = max(ans, cur_len);  // update answer one last time
+    return ans;
+}
+
+void tests() {
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:71]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void tests() {
```

File: strings/longest_substr_without_repeating_chars.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+             << result << '\n';
+
+        if (result == "(FAILED)")
+            exit(0);
+    }
+    cout << "\n********************** Hardcoded TestCases Passed "
+            "**********************\n";
+}
+
+int main() { tests(); }
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:105]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
    tests();  // run self-test implementations
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work on this! 🚀
If you need any help, don't hesitate to ask for help here or in our [**Discord community**](https://discord.gg/c7MnfGFGa6)! We are glad to help. 🎉

**Comment by nishant-mahesh** [comment]
Hey again @Panquesito7, thanks for listing out all the changes I should make. 

Quick follow up -- am I just meant to make all the changes you listed on my local editor, commit the changes and submit another pull request, or is there a way to make quick additions and resolve your comments/ modifications as I see them on github?

If these are basic things I should be knowing, it'd be great if you could direct me to some resources that will help me out here!

Thanks again :)

File: strings/longest_substr_without_repeating_chars.cpp (hunk 3 above)
Comment (nishant-mahesh) [reply strings/longest_substr_without_repeating_chars.cpp:26]: What exactly does this mean? Do you mean I must specify that this function takes a string variable named **input** as input?

**Review by nishant-mahesh — COMMENTED** [review]
I've made all the changes requested (except for the one that said "Missing parameter documentation") for the ```int solver(string input)``` function in the form of a commit (1a2bb83).

File: strings/longest_substr_without_repeating_chars.cpp (hunk 3 above)
Comment (Panquesito7) [reply strings/longest_substr_without_repeating_chars.cpp]: > What exactly does this mean? Do you mean I must specify that this function takes a string variable named **input** as input?

You must add documentation for the parameter: what does it do, how does it work, etc..
You can take this code as a reference. I hope that helps. Let us know if you need any more help! 😃

```c++
/**
 * Utility function to print matrix
 * @tparam n number of matrix size
 * @param board matrix where numbers are saved
 */
template <size_t n>
void printSolution(const std::array<std::array<int, n>, n> &board) {
    std::cout << "\n";
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            std::cout << "" << board[i][j] << " ";
        }
        std::cout << "\n";
    }
}
```

**Comment by nishant-mahesh** [comment]
Hey @Panquesito7. I've completed all the checks that you'd requested. Can this be merged now?

**Comment by nishant-mahesh** [comment]
Hi again @Panquesito7. Just wanted to send a gentle reminder to review my latest version. Thank you 😄

File: strings/longest_substr_without_repeating_chars.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
+/**
+ * @file
+ * @brief Greedy algorithm to find length of longest substring without repeating
+ * characters. The following
+ * [resource](https://www.geeksforgeeks.org/length-of-the-longest-substring-without-repeating-characters/)
+ * was referred to
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:6]:
```suggestion
 * @brief Greedy algorithm to find the [length of the longest substring without repeating
 * characters](https://www.geeksforgeeks.org/length-of-the-longest-substring-without-repeating-characters/)
```

File: strings/longest_substr_without_repeating_chars.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ * (amortized)
+ *
+ * @author [Nishant Mahesh](https://github.com/nishant-mahesh)
+ */
+
+#include <algorithm>  // To use the inbuilt 'max' function
+#include <cassert>
+#include <iostream>
+#include <map>
+#include <string>
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:23]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: strings/longest_substr_without_repeating_chars.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            "**********************\n";
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    tests();  // run self-test implementations
+}
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:128]:
```suggestion
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

**Comment by nishant-mahesh** [comment]
Hey @Panquesito7! I have committed the changes you requested in the previous review. Let me know if there's anything else I may have to change! Thanks for the patience :)

File: strings/longest_substr_without_repeating_chars.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+ * (amortized)
+ *
+ * @author [Nishant Mahesh](https://github.com/nishant-mahesh)
+ */
+
+#include <algorithm>  // To use the inbuilt algorithmic functions, eg: 'max'
+#include <iostream>   // For IO Operations
+#include <map>     // Used to efficiently store and retrieve (key, value) pairs
+#include <string>  // For string operations
+
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:22]:
```suggestion
#include <algorithm>  /// for the built-in algorithmic functions
#include <iostream>   /// for IO Operations
#include <map>       /// used to efficiently store and retrieve (key, value) pairs
#include <string>   /// for std::string
```

File: strings/longest_substr_without_repeating_chars.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+using std::cin;
+using std::cout;
+using std::map;
+using std::max;
+using std::pair;
+using std::string;
+
+/**
+ * @param string named "input"
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:31]: Yes, we know it's a string named `input`, but what is the parameter used for? This needs to be documented. :)
Let us know if you need any help with anything!

File: strings/longest_substr_without_repeating_chars.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    cout << "\n********************** Hardcoded TestCases Passed "
+            "**********************\n";
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    tests();  // run self-test implementations
```
Comment (Panquesito7) [inline strings/longest_substr_without_repeating_chars.cpp:125]:
```suggestion
    tests();  // run self-test implementations
    return 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by nishant-mahesh** [comment]
Hello again @Panquesito7! I have made the changes based on your last code review. Let me know if you'd like any other changes to be made :)

**Comment by nishant-mahesh** [comment]
Hey @Panquesito7 . It's just the ```assert``` usage that's left to review right now. Have a look at my comment on why I've used an alternative to ```assert``` and let me know what you think!

**Comment by nishant-mahesh** [comment]
Hello @Panquesito7. Gentle reminder to check this out again :)

**Comment by nishant-mahesh** [comment]
Hey @Panquesito7. Gentle reminder again, to have a look at this. Thanks!

**Comment by nishant-mahesh** [comment]
Hi @Panquesito7. Could you please review these changes? I'd appreciate it greatly :)

**Pull request closed without merging**